File: examples/nodl_tutorials/navigate_to_pose_nodl/src/navigation.cpp
// SPDX-FileCopyrightText: 2026 Open Source Robotics Foundation, Inc.
// SPDX-License-Identifier: Apache-2.0

#include "navigation.hpp"  // NOLINT(build/include_subdir)

#include <chrono>
#include <string>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"

using std::chrono_literals::operator""s;

int run_navigation(
  int argc, char * argv[], const rclcpp::Node::SharedPtr & node, const NavigateToPoseActionClient::SharedPtr & client)
{
  const std::vector<std::string> args = rclcpp::remove_ros_arguments(argc, argv);
  if (args.size() == 1) {
    // Keep the interface available for the headless conformance test.
    rclcpp::spin(node);
    return 0;
  }
  if (args.size() != 3) {
    RCLCPP_ERROR(node->get_logger(), "usage: %s [x y]", args.front().c_str());
    return 2;
  }

  double x;
  double y;
  try {
    x = std::stod(args[1]);
    y = std::stod(args[2]);
  } catch (const std::exception & error) {
    RCLCPP_ERROR(node->get_logger(), "invalid coordinates: %s", error.what());
    return 2;
  }

  if (!client->wait_for_action_server(10s)) {
    RCLCPP_ERROR(node->get_logger(), "NavigateToPose server was not available");
    return 1;
  }

  nav2_msgs::action::NavigateToPose::Goal goal;
  goal.pose.header.frame_id = "map";
  goal.pose.header.stamp = node->get_clock()->now();
  goal.pose.pose.position.x = x;
  goal.pose.pose.position.y = y;
  goal.pose.pose.orientation.w = 1.0;

  auto goal_future = client->async_send_goal(goal);
  if (rclcpp::spin_until_future_complete(node, goal_future) != rclcpp::FutureReturnCode::SUCCESS) {
    RCLCPP_ERROR(node->get_logger(), "Navigation goal could not be sent");
    return 1;
  }
  const auto goal_handle = goal_future.get();
  if (!goal_handle) {
    RCLCPP_ERROR(node->get_logger(), "Navigation goal was rejected");
    return 1;
  }

  RCLCPP_INFO(node->get_logger(), "Navigating to (%.2f, %.2f)", x, y);
  auto result_future = client->async_get_result(goal_handle);
  if (rclcpp::spin_until_future_complete(node, result_future) != rclcpp::FutureReturnCode::SUCCESS) {
    RCLCPP_ERROR(node->get_logger(), "Navigation result was not available");
    return 1;
  }

  const bool succeeded = result_future.get().code == rclcpp_action::ResultCode::SUCCEEDED;
  RCLCPP_INFO(node->get_logger(), "Navigation %s", succeeded ? "succeeded" : "failed");
  return succeeded ? 0 : 1;
}

File: examples/nodl_tutorials/navigate_to_pose_nodl/navigate_to_pose_nodl/navigation.py
# SPDX-FileCopyrightText: 2026 Open Source Robotics Foundation, Inc.
# SPDX-License-Identifier: Apache-2.0
"""Behavior shared by the conventional and NoDL-backed navigation clients."""

import argparse
import sys

import rclpy
from action_msgs.msg import GoalStatus
from nav2_msgs.action import NavigateToPose
from rclpy.utilities import remove_ros_args


def goal_coordinates(args=None):
    """Read a map-frame target while leaving ROS arguments for rclpy."""
    command = sys.argv if args is None else [sys.argv[0], *args]
    parser = argparse.ArgumentParser(description='Send one Nav2 NavigateToPose goal.')
    parser.add_argument('x', type=float, nargs='?', help='Goal x coordinate in the map frame.')
    parser.add_argument('y', type=float, nargs='?', help='Goal y coordinate in the map frame.')
    coordinates = parser.parse_args(remove_ros_args(args=command)[1:])
    if (coordinates.x is None) != (coordinates.y is None):
        parser.error('provide both x and y, or neither')
    return coordinates


def navigate(node, x, y):
    """Send one goal through the action client supplied by the node."""
    client = node.action_cli_navigate_to_pose
    if not client.wait_for_server(timeout_sec=10.0):
        node.get_logger().error('NavigateToPose server was not available')
        return False

    goal = NavigateToPose.Goal()
    goal.pose.header.frame_id = 'map'
    goal.pose.header.stamp = node.get_clock().now().to_msg()
    goal.pose.pose.position.x = x
    goal.pose.pose.position.y = y
    goal.pose.pose.orientation.w = 1.0

    goal_future = client.send_goal_async(goal)
    rclpy.spin_until_future_complete(node, goal_future)
    goal_handle = goal_future.result()
    if goal_handle is None or not goal_handle.accepted:
        node.get_logger().error('Navigation goal was rejected')
        return False

    node.get_logger().info(f'Navigating to ({x:.2f}, {y:.2f})')
    result_future = goal_handle.get_result_async()
    rclpy.spin_until_future_complete(node, result_future)
    result = result_future.result()
    succeeded = result is not None and result.status == GoalStatus.STATUS_SUCCEEDED
    node.get_logger().info('Navigation succeeded' if succeeded else 'Navigation failed')
    return succeeded


def run(client_type, args=None):
    """Run either client implementation with the same navigation behavior."""
    coordinates = goal_coordinates(args)
    rclpy.init(args=args)
    node = client_type()
    try:
        # With no goal, keep the interface available for the headless conformance test.
        if coordinates.x is None:
            try:
                rclpy.spin(node)
            except KeyboardInterrupt:
                pass
            return 0
        return 0 if navigate(node, coordinates.x, coordinates.y) else 1
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
